akka-http-core/src/main/scala/akka/http/impl/engine/ws/Handshake.scala

```diff
@@ -144,15 +145,20 @@ private[http] object Handshake {
           concatenated value to obtain a 20-byte value and base64-
           encoding (see Section 4 of [RFC4648]) this 20-byte hash.
     */
-    def buildResponse(key: `Sec-WebSocket-Key`, handler: Either[Graph[FlowShape[FrameEvent, FrameEvent], Any], Graph[FlowShape[Message, Message], Any]], subprotocol: Option[String]): HttpResponse =
+    def buildResponse(key: `Sec-WebSocket-Key`, handler: Either[Graph[FlowShape[FrameEvent, FrameEvent], Any], Graph[FlowShape[Message, Message], Any]], subprotocol: Option[String], requestedProtocols: immutable.Seq[String]): HttpResponse =
       HttpResponse(
         StatusCodes.SwitchingProtocols,
-        subprotocol.map(p ⇒ `Sec-WebSocket-Protocol`(Seq(p))).toList :::
+        selectSubProtocol(requestedProtocols, subprotocol).toList :::
           List(
             UpgradeHeader,
             ConnectionUpgradeHeader,
             `Sec-WebSocket-Accept`.forKey(key),
             UpgradeToWebSocketResponseHeader(handler)))
+
+    // subprotocol has been validated to be either * or in the list
+    private[akka] def selectSubProtocol(requestedProtocols: immutable.Seq[String], subprotocol: Option[String]): Option[`Sec-WebSocket-Protocol`] = {
+      subprotocol.map(sp ⇒ `Sec-WebSocket-Protocol`(Seq(if (SubProtocolWildcard.eq(sp)) requestedProtocols.head else sp)))
```

Comment:
Why not `sp == SubProtocolWildcard`? Weems obscure to rely on object equality here since it'd assume the `"*"` passed in by e.g. Play is interned (or refers to this constant)?